plugins/Model_GAN/Trainer.py

```diff
@@ -24,130 +29,216 @@ def __init__(self, model, fn_A, fn_B, batch_size):
+        # ==========  
+        loss_G += K.mean(K.abs(fake_rgb - real))
+        # ==========
+        
+        # Edge loss (similar with total variation loss) 
+        loss_G += 1 * K.mean(K.abs(self.first_order(fake_rgb, axis=1) - self.first_order(real, axis=1)))
+        loss_G += 1 * K.mean(K.abs(self.first_order(fake_rgb, axis=2) - self.first_order(real, axis=2)))
+        
+        
+        # Perceptual Loss
+        if not vggface_feat is None:
+            def preprocess_vggface(x):
+                x = (x + 1)/2 * 255 # channel order: BGR
+                #x[..., 0] -= 93.5940
+                #x[..., 1] -= 104.7624
+                #x[..., 2] -= 129.
+                x -= [91.4953, 103.8827, 131.0912]
+                return x
+            pl_params = (0.011, 0.11, 0.1919)
+            real_sz224 = tf.image.resize_images(real, [224, 224])
+            real_sz224 = Lambda(preprocess_vggface)(real_sz224)
+            # ==========
+            fake_sz224 = tf.image.resize_images(fake_rgb, [224, 224]) 
+            fake_sz224 = Lambda(preprocess_vggface)(fake_sz224)
+            # ==========   
+            real_feat55, real_feat28, real_feat7 = vggface_feat(real_sz224)
+            fake_feat55, fake_feat28, fake_feat7  = vggface_feat(fake_sz224)    
+            loss_G += pl_params[0] * K.mean(K.abs(fake_feat7 - real_feat7))
+            loss_G += pl_params[1] * K.mean(K.abs(fake_feat28 - real_feat28))
+            loss_G += pl_params[2] * K.mean(K.abs(fake_feat55 - real_feat55))
+        
+        return loss_D, loss_G
     
     def show_sample(self, display_fn):
         _, wA, tA = next(self.train_batchA)
         _, wB, tB = next(self.train_batchB)
-        self.showG(tA, tB, display_fn)
-
-    def showG(self, test_A, test_B, display_fn):
-        def display_fig(name, figure_A, figure_B):
-            figure = np.concatenate([figure_A, figure_B], axis=0 )
-            columns = 4
-            elements = figure.shape[0]
-            figure = figure.reshape((columns,(elements//columns)) + figure.shape[1:])
-            figure = stack_images(figure)
-            figure = np.clip((figure + 1) * 255 / 2, 0, 255).astype('uint8')
-            display_fn(figure, name)
-
-        out_test_A_netGA = self.model.netGA.predict(test_A)
-        out_test_A_netGB = self.model.netGB.predict(test_A)
-        out_test_B_netGA = self.model.netGA.predict(test_B)
-        out_test_B_netGB = self.model.netGB.predict(test_B)
-
-        figure_A = np.stack([
-            test_A,
-            out_test_A_netGA[1],
-            out_test_A_netGB[1],
-            ], axis=1 )
-        figure_B = np.stack([
-            test_B,
-            out_test_B_netGB[1],
-            out_test_B_netGA[1],
-            ], axis=1 )
-        
-        display_fig("raw", figure_A, figure_B)       
-
+        display_fn(self.showG(tA, tB, self.path_A, self.path_B), "raw")
+        display_fn(self.showG(wA, wB, self.path_bgr_A, self.path_bgr_B), "masked")

```

Comment:
I believe this is the reason there's warped masked output. Not sure why @Clorr used the warped version here (`wA` instead of `tA`)